core/src/main/java/io/grpc/internal/RetriableStream.java

```diff
@@ -225,9 +225,12 @@ private void commitAndRun(Substream winningSubstream) {
   @Nullable // returns null when cancelled
   private Substream createSubstream(int previousAttemptCount, boolean isTransparentRetry) {
     // increment only when >= 0, i.e. not cancelled
-    if (inFlightSubStreams.updateAndGet(value -> value < 0 ? value : value + 1) < 0) {
-      return null;
-    }
+    int inFlight = inFlightSubStreams.get();
+    do {
```

Comment:
shall we move inFlightSubStreams.get() inside do loop?

Reply:
Absolutely. Thank you. That would have been bad.

That's what I get for writing 6 lines of code over several hours time.